### PR TITLE
Move cascading delete to onAfterDelete for SiteTree

### PR DESCRIPTION
**Pull request by unclecheese**
Resolves https://github.com/silverstripe/silverstripe-framework/issues/5014#issuecomment-335975688

### CODE REVIEW

File: tests/php/Model/SiteTreeTest.php [hunk 1]
```diff
@@ -332,6 +332,26 @@ public function testRestoreToStage()
...
+        $fail = true;
+        try {
+            $obj->delete();
+        } catch (\LogicException $e) {
+            $fail = false;
+        }
+        if ($fail) {
+            $this->fail('Failed to throw delete exception');
+        }
+        $this->assertEquals($count, SiteTree::get()->count());
```
Comment (dhensby) [inline tests/php/Model/SiteTreeTest.php]: `assertCount` please

File: tests/php/Model/SiteTreeTest.php [hunk 2]
```diff
@@ -332,6 +332,26 @@ public function testRestoreToStage()
         $this->assertInstanceOf('Page', $requeriedPage);
     }
 
+    public function testNoCascadingDeleteWithoutID()
+    {
+        Config::nest();
```
Comment (dhensby) [inline tests/php/Model/SiteTreeTest.php]: `Config` nesting is not needed as the test suite does this itself

File: code/Model/SiteTree.php [hunk 3]
```diff
@@ -1481,9 +1481,10 @@ public function syncLinkTracking()
         $this->extend('augmentSyncLinkTracking');
     }
 
-    public function onBeforeDelete()
+    public function onAfterDelete()
```
Comment (tractorcow) [inline code/Model/SiteTree.php]: As commented on #1988 

Moving the cascade delete into the "after" introduces a rather concerning shift in integrity I must point out. I can't see how this is necessary in order to simply avoid deleting all records, when it seems all that's necessary is an `isInDB()` check on children enumeration.

If you delete parents first, but fail to delete a child, you end up with your application in an inconsistent state. It needs to be leaf-first not node-first to ensure this consistency.

**Comment by unclecheese** [comment]
Closing as the overall approach violates data integrity. The only change needed to fix this is a change to hierarchy.

**Pull request closed without merging**